zipline/data/fx/in_memory.py

```diff
@@ -0,0 +1,70 @@
+"""Interface and definitions for foreign exchange rate readers.
+"""
+from interface import implements
+
+from .base import FXRateReader
+
+
+class InMemoryFXRateReader(implements(FXRateReader)):
+    """
+    A simple in-memory FXRateReader.
+
+    This is primarily used for testing.
+
+    Parameters
+    ----------
+    data : dict
+        Nested map from rate name -> quote currency -> pd.DataFrame
+        Leaf frames should be indexed by (dates, base currencies).
+    default_rate : str
+        Rate to use when ``get_rates`` is called with a rate of 'default'.
+    """
+
+    def __init__(self, data, default_rate):
+        self._data = data
+        self._default_rate = default_rate
+
+    def get_rates(self, rate, quote, bases, dts):
+        """Get rates to convert ``bases`` into ``quote``.
+        """
+        if rate == 'default':
+            rate = self._default_rate
+
+        df = self._data[rate][quote]
```

Comment:
Looks like calling `.get_rates(bogus_rate, valid_quote ...)` returns the exact same KeyError from the exact same line as `.get_rates(valid_rate, bogus_quote, ...)`. Is it worth splitting this line into two lines, to help distinguish them? (Maybe even using different KeyError subclasses for the two? Seems like that would encourage weird control flow in calling code, though.)